### PR TITLE
8304939: os::win32::exit_process_or_thread should be marked noreturn

### PR DESCRIPTION
**Pull request by TheShermanTanker**
On Windows we have a non-trivial function (exit_process_or_thread) that provides the implementation of various functions like os::die, os::abort, &etc. Those os functions are marked as noreturn, so this implementation helper should also be noreturn.

The current change does several things around exit_process_or_thread, which is moved out of the os::win32 class since it does not need access to anything in that class other than the Ept enum, which is also moved to os_windows.cpp as well. The signature is changed to void, since exit_process_or_thread's current return value is simply the exit code parameter that it is passed, and to qualify as noreturn it should not return its current value of int. The only usage of this return value is in thread_native_entry, and this usage can easily be replaced by returning the res local that exit_process_or_thread is passed. For this os::infinite_sleep takes the place of the return statement in exit_process_or_thread, to make sure it qualifies as noreturn.

Although not mentioned in the title, raise_fail_fast has also been fixed. RaiseFailFastException is not itself noreturn, so os::infinite_sleep has been added there as well, to make raise_fail_fast qualify as a noreturn method

All the changes mentioned above fixes any Undefined Behaviour on Windows with regards to noreturn, by making all code marked noreturn (os::die() os::exit() etc) never return. This is not a problem on other platforms since their implementations of os::abort or os::die etc already call noreturn methods.

This fix also fixes compilation on gcc windows, which warns about noreturn methods returning prior to this changeset

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8304939](https://bugs.openjdk.org/browse/JDK-8304939): os::win32::exit_process_or_thread should be marked noreturn (**Enhancement** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16303/head:pull/16303` \
`$ git checkout pull/16303`

Update a local copy of the PR: \
`$ git checkout pull/16303` \
`$ git pull https://git.openjdk.org/jdk.git pull/16303/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16303`

View PR using the GUI difftool: \
`$ git pr show -t 16303`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16303.diff">https://git.openjdk.org/jdk/pull/16303.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16303#issuecomment-1774549611)

### CODE REVIEW

**Comment by TheShermanTanker** [comment]
Bumping, anyone?

**Comment by dholmes-ora** [comment]
On my to-do list ...

**Comment by dholmes-ora** [comment]
> moved out of the os::win32

This doesn't seem to add any value. Seemed fine to me in os::win32 versus os class.

File: src/hotspot/os/windows/os_windows.cpp
Comment (dholmes-ora) [inline src/hotspot/os/windows/os_windows.cpp:3935]: Again there needs to be a very strong justification for doing it this way.

File: src/hotspot/os/windows/os_windows.cpp
Comment (dholmes-ora) [inline src/hotspot/os/windows/os_windows.cpp:4118]: exit_code was clearer and avoids changes below

File: src/hotspot/os/windows/vmError_windows.cpp
Comment (dholmes-ora) [inline src/hotspot/os/windows/vmError_windows.cpp:71]: Curious that MS do not declare this as no_return themselves.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Supporting no_return seems a reasonable thing to do.

General complaint is that there is too much incidental change in the PR that to me adds no value. 

Some specific objections below.

Thanks

File: src/hotspot/os/windows/os_windows.cpp
Comment (TheShermanTanker) [reply src/hotspot/os/windows/os_windows.cpp:4118]: Noted, will return the name to exit_code

File: src/hotspot/os/posix/os_posix.cpp
Comment (TheShermanTanker) [reply src/hotspot/os/posix/os_posix.cpp:894]: ah, when I was searching for functions to implement the noreturn with and stumbled across this one, I thought it could do with a goto instead of a while true since the intent seemed to be clearer. I can revert this if need be

**Comment by TheShermanTanker** [comment]
> > moved out of the os::win32
> 
> This doesn't seem to add any value. Seemed fine to me in os::win32 versus os class.

I noticed that the method wasn't used anywhere outside of os_windows.cpp, and since I was rewriting it to be noreturn, I figured that a little refactoring of scope (from os::win32 to os_windows.cpp file scope) could help here

File: src/hotspot/os/posix/os_posix.cpp
Comment (dholmes-ora) [reply src/hotspot/os/posix/os_posix.cpp:894]: Please revert - `while (true)` conveys the exact intent.

**Comment by dholmes-ora** [comment]
> I figured that a little refactoring of scope (from os::win32 to os_windows.cpp file scope) could help here

The very loose, not well followed, historical convention here is that the Windows specific os class contains the methods defined by os.hpp, while the implementation details go into the os::win32 class. In many cases the choice is somewhat arbitrary, but there should still be a good reason to move something around. Unnecessary refactoring just makes the PR harder to understand.

**Comment by TheShermanTanker** [comment]
Addressed some of the review comments

Side note: Should the Style Guide only permit noreturn for void methods? It's Undefined Behaviour when applied to something that returns int for instance, such as exit_process_or_thread here (which I had to refactor to void)

**Comment by dholmes-ora** [comment]
> Side note: Should the Style Guide only permit noreturn for void methods? It's Undefined Behaviour when applied to something that returns int for instance, such as exit_process_or_thread here (which I had to refactor to void)

I think it is implied that attributes should only be used in a way that is valid.

File: src/hotspot/os/windows/os_windows.cpp
Comment (dholmes-ora) [inline src/hotspot/os/windows/os_windows.cpp:511]: Nit: code -> exit_code

**Review by dholmes-ora — APPROVED** [review]
I think the refactoring could have been rolled back further, but okay.

Needs second review.

Thanks

**Comment by TheShermanTanker** [comment]
Paging for @kimbarrett, since the original issue was created by him

**Review by kimbarrett — CHANGES_REQUESTED** [review]

File: src/hotspot/os/windows/os_windows.cpp
Comment (dholmes-ora) [reply src/hotspot/os/windows/os_windows.cpp:571]: It has to call `exit_process_or_thread` because of the exit bug. That means we will directly call `_endthreadex` instead of implicitly doing that by having the entry function return. The `return res` is redundant but keeps the compiler happy given the required declaration of the entry method.

**Review by kimbarrett — APPROVED** [review]
To speed along the process, I'm approving this even though I still want the pointless
forward declaration of `thread_native_entry` removed.  I'll trust you to do that without
needing a re-review by me.

**Comment by TheShermanTanker** [comment]
> To speed along the process, I'm approving this even though I still want the pointless forward declaration of `thread_native_entry` removed. I'll trust you to do that without needing a re-review by me.

I'll do just that, but I have a couple of questions: What about the comment accompanying the declaration? And is os::infinite_sleep appropriate for this purpose? I couldn't seem to find anything else to make exit_process_or_thread nor raise_fail_fast noreturn qualified

File: src/hotspot/os/windows/os_windows.cpp
Comment (TheShermanTanker) [reply src/hotspot/os/windows/os_windows.cpp:515]: I understand, but what about the useful (at least to me) comment? Should I move it to the definition of the method?

File: src/hotspot/os/windows/os_windows.cpp
Comment (TheShermanTanker) [reply src/hotspot/os/windows/os_windows.cpp:571]: I'm not too sure what to make of this, since I don't know what the exit bug is about (Also, the return res doesn't cause an issue on MSVC under any circumstance, and would only do so on gcc if thread_native_entry was marked noreturn, which it isn't). I simply changed the return value to keep the original semantics of the code unchanged, I guess I should take this to mean I should keep my current changes as is?

**Comment by TheShermanTanker** [comment]
> > I figured that a little refactoring of scope (from os::win32 to os_windows.cpp file scope) could help here
> 
> The very loose, not well followed, historical convention here is that the Windows specific os class contains the methods defined by os.hpp, while the implementation details go into the os::win32 class. In many cases the choice is somewhat arbitrary, but there should still be a good reason to move something around. Unnecessary refactoring just makes the PR harder to understand.

Just noticed something in this review comment - I didn't move the method into the os class, I moved it into os_windows.cpp file scope

File: src/hotspot/os/windows/os_windows.cpp
Comment (kimbarrett) [reply src/hotspot/os/windows/os_windows.cpp:515]: Sure, move the comment.

File: src/hotspot/os/windows/os_windows.cpp
Comment (kimbarrett) [reply src/hotspot/os/windows/os_windows.cpp:571]: Yes, keep as is in this area.

**Comment by TheShermanTanker** [comment]
@dholmes-ora @kimbarrett Please see if you are fine with the final set of changes

**Review by dholmes-ora — APPROVED** [review]
Still seems fine to me. Thanks.

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Comment by TheShermanTanker** [comment]
Thanks for the reviews!

/integrate

**Pull request closed without merging**
